lib/geoengineer/utils/aws_clients.rb

```diff
@@ -11,80 +11,168 @@ def self.stubbed?
     @stub_aws || false
   end
 
-  # Clients
-  def self.cloudwatch
-    @aws_cloudwatch ||= Aws::CloudWatch::Client.new({ stub_responses: stubbed? })
-  end
-
-  def self.cloudwatchevents
-    @aws_cloudwatchevents ||= Aws::CloudWatchEvents::Client.new({ stub_responses: stubbed? })
-  end
-
-  def self.dynamo
-    @aws_dynamo ||= Aws::DynamoDB::Client.new({ stub_responses: stubbed? })
-  end
-
-  def self.ec2
-    @aws_ec2 ||= Aws::EC2::Client.new({ stub_responses: stubbed? })
-  end
-
-  def self.elasticache
-    @aws_elasticache ||= Aws::ElastiCache::Client.new({ stub_responses: stubbed? })
-  end
-
-  def self.elasticsearch
-    @aws_elasticsearch ||= Aws::ElasticsearchService::Client.new({ stub_responses: stubbed? })
+  def self.client_params(provider = nil)
+    client_params = { stub_responses: stubbed? }
+    client_params[:region] = provider.region if provider
+    client_params
   end
 
-  def self.elb
-    @aws_elb ||= Aws::ElasticLoadBalancing::Client.new({ stub_responses: stubbed? })
+  def self.client_cache(client_name, provider, client)
+    @client_cache ||= {}
+    key = "#{client_name}_" + (provider&.terraform_id || GeoEngineer::Resource::DEFAULT_PROVIDER)
+    @client_cache[key] ||= client
   end
 
-  def self.iam
-    @aws_iam ||= Aws::IAM::Client.new({ stub_responses: stubbed? })
+  # Clients
+  def self.cloudwatch(provider = nil)
+    self.client_cache(
+      'cloudwatch',
+      provider,
+      Aws::CloudWatch::Client.new(client_params(provider))
+    )
   end
-
-  def self.kinesis
-    @aws_kinesis ||= Aws::Kinesis::Client.new({ stub_responses: stubbed? })
+
+  def self.cloudwatchevents(provider = nil)
+    self.client_cache(
+      'cloudwatchevents',
+      provider,
+      Aws::CloudWatchEvents::Client.new(client_params(provider))
```

Comment:
maybe passing the client constant and then initializing inside the client_cahe method would be best